torch/csrc/cuda/comm.cpp

```diff
@@ -47,7 +47,13 @@ std::vector<Tensor> broadcast(const Tensor& tensor, IntList devices) {
   {
 #endif
     auto & gpu_type = type.toBackend(type.is_sparse() ? at::kSparseCUDA : at::kCUDA);
-    for (auto device : devices) {
+    if (type.is_cuda()) {
+      tensors.push_back(tensor);
+    } else {
+      AutoGPU _gpu_guard(devices[0]);
+      tensors.push_back(gpu_type.copy(tensor, true));
+    }
```

Comment:
Can we just make it this? It will let us avoid duplicating the body.
```cpp
IntList loopDevices = ... ? devices : devices.slice(1);
for (auto device : loopDevices) {
  ...
}
```

Reply:
Also, a minor improvement would be to move the `AutoGPU` outside of the loop, and only call `setDevice` inside (less calls to CUDA API)